src/src/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -0,0 +1,317 @@
+     * @throws AuthenticationException 
+     * @throws IllegalArgumentException If {@link AuthenticationResult} is null. 
+     */
+    void updateCachedItemWithResult(final String resource, final String clientId, final AuthenticationResult result, 
+            final TokenCacheItem cachedItem) throws AuthenticationException {
+        if (result == null) {
+            Logger.v(TAG, "AuthenticationResult is not, cannot update cache.");
+            throw new IllegalArgumentException("result");
+        }
+        
+        if (result.getStatus() == AuthenticationStatus.Succeeded) {
+            Logger.v(TAG, "Save returned AuthenticationResult into cache.");
+            if (cachedItem != null && cachedItem.getUserInfo() != null && result.getUserInfo() == null) {
+                result.setUserInfo(cachedItem.getUserInfo());
+                result.setIdToken(cachedItem.getRawIdToken());
+                result.setTenantId(cachedItem.getTenantId());
+            }
+            
+            updateTokenCache(resource, clientId, result);
+        } else if (AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT.equalsIgnoreCase(result.getErrorCode())) {
+            // remove Item if oauth2_error is invalid_grant
+            Logger.v(TAG, "Received INVALID_GRANT error code, remove existing cache entry.");
+            removeTokenCacheItem(cachedItem, resource);
+        }
+    }
+    
+    /**
+     * Update token cache with returned auth result. 
+     * @param result
+     */
+    void updateTokenCache(final String resource, final String clientId, final AuthenticationResult result) {
+        if (result == null || StringExtensions.IsNullOrBlank(result.getAccessToken())) {
+            return;
+        }
+        
+        if (result.getUserInfo() != null) {
+            // update cache entry with displayableId
+            if (!StringExtensions.IsNullOrBlank(result.getUserInfo().getDisplayableId())) {
+                setItemToCacheForUser(resource, clientId, result, result.getUserInfo().getDisplayableId());
+            }
+            
+            // update cache entry with userId
+            if (!StringExtensions.IsNullOrBlank(result.getUserInfo().getUserId())) {
+                setItemToCacheForUser(resource, clientId, result, result.getUserInfo().getUserId());
+            }
+        }
+        
+        // update for empty userid
+        setItemToCacheForUser(resource, clientId, result, null);
+    }
+    
+    /**
+     * Remove token from cache.
+     * {@link RefreshItem#mKeysWithUser} is holding a list of keys related to user for removal. 
+     * 1) If refresh with resource specific token cache entry, clear RT with key(R,C,U,A)
+     * 2) If refresh with MRRT, clear RT (C,U,A) and (R,C,U,A)
+     * 3) if refresh with FRT, clear RT with (U,A) 
+     * @throws AuthenticationException 
+     */
+    void removeTokenCacheItem(final TokenCacheItem tokenCacheItem, final String resource) 
```

Comment:
oh yea, good call, updated. 

---
In reply to: [63414464](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63414464) [](ancestors = 63414464)